ipu-plugin/pkg/p4rtclient/p4rtclient.go

```diff
@@ -68,22 +69,24 @@ func checkMacAddresses(macAddresses ...string) ([]byte, error) {
 func (p *p4rtclient) ProgramFXPP4Rules(ruleSets []types.FxpRuleBuilder) error {
 	for _, r := range ruleSets {
 		p4rule := []string{r.Action, r.P4br, r.Control, r.Metadata}
-		err := utils.RunP4rtCtlCommand(p.p4rtBin, p.p4rtIpPort, p4rule...)
+		stdout, err := utils.RunP4rtCtlCommand(p.p4rtBin, p.p4rtIpPort, p4rule...)
 		if err != nil {
-			log.Info("WARNING: Failed to program p4rule: ", p4rule)
-		}
+                       if strings.Contains(stdout, "ALREADY_EXISTS") {
+			       log.Info("p4rule already exists. Delete and reprogram: ", p4rule)
+			       metaPrefix := strings.Index(r.Metadata, ",action")
+			       p4DelR := types.FxpRuleBuilder{Action:"del-entry", P4br:r.P4br, Control:r.Control, Metadata:r.Metadata[:metaPrefix]}
+			       p4DelRuleStr := []string{p4DelR.Action,p4DelR.P4br, p4DelR.Control, p4DelR.Metadata}
+			       // Delete the existing rule first
+			       utils.RunP4rtCtlCommand(p.p4rtBin, p.p4rtIpPort, p4DelRuleStr...)
+			       time.Sleep(1)
+			       // Now program the new rule
+			       utils.RunP4rtCtlCommand(p.p4rtBin, p.p4rtIpPort, p4rule...)
+		       }
```

Comment:
Instead of delete and then add, should we do a modify instead?

Reply:
We don't have a modify option in p4rt-ctl command. But, I did use modify option with OPCODE way of program rules. But there was some concern which it was not a good fit for all the use cases (couldn't recall what it was exactly).